nomad/deploymentwatcher/deployment_watcher.go

```diff
@@ -702,6 +695,33 @@ func (w *deploymentWatcher) shouldFail() (fail, rollback bool, err error) {
 	return fail, false, nil
 }
 
+func (w *deploymentWatcher) shouldFailEarly(deployment *structs.Deployment, alloc *structs.AllocListStub, dstate *structs.DeploymentState) bool {
+	if !alloc.DeploymentStatus.IsUnhealthy() {
+		return false
+	}
+
+	hasProgressDeadline := dstate.ProgressDeadline != 0
+
+	// Fail on the first unhealthy allocation if no progress deadline is specified.
+	if !hasProgressDeadline {
+		w.logger.Debug("failing deployment because an allocation failed and the deployment is not progress based", alloc.ID)
+		return true
+	}
+
+	if deployment.Active() {
+		reschedulePolicy := w.j.LookupTaskGroup(alloc.TaskGroup).ReschedulePolicy
+		isRescheduleEligible := alloc.RescheduleEligible(reschedulePolicy, time.Now())
+		if !isRescheduleEligible {
+			// We have run out of reschedule attempts: do not wait for the progress deadline to expire because
+			// we know that we will not be able to try to get another allocation healthy
+			w.logger.Debug("failing deployment because an allocation has failed and the task group has run out of reschedule attempts", alloc.ID)
```

Comment:
hclog requires pairs of key-value pairs following the first param.

```suggestion
			w.logger.Debug("failing deployment because an allocation has failed and the task group has run out of reschedule attempts", "alloc", alloc.ID)
```

Otherwise you end up with logs like this:

> 2023-06-12T11:47:52.630-0400 [DEBUG] nomad.deployments_watcher: failing deployment because an allocation has failed and the task group has run out of reschedule attempts: deployment_id=d73fc44f-d9a9-583a-d563-13a758dca813 job="<ns: \"default\", id: \"mock\">" EXTRA_VALUE_AT_END=a947bcfa-81b4-2e81-8865-d92474fcdf87

Reply:
🤦 Nice catch, thanks for that